processing/src/main/java/io/druid/query/ordering/StringComparators.java

```diff
@@ -340,18 +345,81 @@ public String toString()
     {
       return StringComparators.STRLEN_NAME;
     }
+
+    @Override
+    public byte[] getCacheKey()
+    {
+      return new byte[]{(byte) STRLEN_CACHE_ID};
+    }
   }
 
-  public static StringComparator makeComparator(String type)
+  private static BigDecimal convertStringToBigDecimal(String input) {
+    // treat unparseable Strings as nulls
+    BigDecimal bd = null;
+    try {
+      bd = new BigDecimal(input);
+    } catch (NullPointerException | NumberFormatException ex) {
+    }
+    return bd;
+  }
+
+  public static class NumericComparator extends StringComparator
   {
-    if (type.equals(StringComparators.LEXICOGRAPHIC_NAME)) {
-      return LEXICOGRAPHIC;
-    } else if (type.equals(StringComparators.ALPHANUMERIC_NAME)) {
-      return ALPHANUMERIC;
-    } else if (type.equals(StringComparators.STRLEN_NAME)) {
-      return STRLEN;
-    } else {
-      throw new IAE("Unknown string comparator[%s]", type);
+    @Override
+    public int compare(String o1, String o2)
+    {
+      if (o1 == o2) {
+        return 0;
+      }
+
+      // Creating a BigDecimal from a String is expensive (involves copying the String into a char[])
+      // Converting the String to a Long first is faster.
+      // We optimize here with the assumption that integer values are more common than floating point.
+      Long long1 = o1 == null ? null : Longs.tryParse(o1);
+      Long long2 = o2 == null ? null : Longs.tryParse(o2);
+
+      final BigDecimal bd1 = long1 == null ? convertStringToBigDecimal(o1) : new BigDecimal(long1);
```

Comment:
also added some null checks to the NumericComparator before the attempted numeric conversions